be/src/exec/vectorized/hash_joiner.h

```diff
@@ -157,12 +209,72 @@ class HashJoiner final : public pipeline::ContextWithDependency {
         // Post probe needn't process _other_join_conjunct_ctxs, because they
         // are `ON` predicates, which need to be processed only on probe phase.
         if (chunk && !chunk->is_empty() && !_conjunct_ctxs.empty()) {
+            // TODO(satanson): _conjunct_ctxs should including local runtime in-filters.
             ExecNode::eval_conjuncts(_conjunct_ctxs, chunk.get());
         }
     }
 
     static std::string _get_join_type_str(TJoinOp::type join_type);
 
+    Status _create_runtime_in_filters(RuntimeState* state) {
+        SCOPED_TIMER(_build_runtime_filter_timer);
+
+        if (_ht.get_row_count() > 1024) {
+            return Status::OK();
+        }
+
+        if (_ht.get_row_count() > 0) {
+            // there is a bug (DSDB-3860) in old planner if probe_expr is not slot-ref, and this fix is workaround.
+            size_t size = _build_expr_ctxs.size();
+            std::vector<bool> to_build(size, true);
+            for (int i = 0; i < size; i++) {
+                ExprContext* expr_ctx = _probe_expr_ctxs[i];
+                to_build[i] = (expr_ctx->root()->is_slotref());
+            }
+
+            for (size_t i = 0; i < size; i++) {
+                if (!to_build[i]) continue;
+                ColumnPtr column = _ht.get_key_columns()[i];
+                Expr* probe_expr = _probe_expr_ctxs[i]->root();
+                // create and fill runtime IN filter.
+                // create and fill runtime IN filter.
+                VectorizedInConstPredicateBuilder builder(state, _pool, probe_expr);
+                builder.set_eq_null(_is_null_safes[i]);
+                builder.use_as_join_runtime_filter();
+                Status st = builder.create();
+                if (!st.ok()) continue;
+                builder.add_values(column, kHashJoinKeyColumnOffset);
+                _runtime_in_filters.push_back(builder.get_in_const_predicate());
+            }
+        }
+
+        COUNTER_UPDATE(_runtime_filter_num, static_cast<int64_t>(_runtime_in_filters.size()));
+        return Status::OK();
+    }
+
+    Status _create_runtime_bloom_filters(RuntimeState* state, int64_t limit) {
+        for (auto* rf_desc : _build_runtime_filters) {
+            rf_desc->set_is_pipeline(true);
+            // skip if it does not have consumer.
+            if (!rf_desc->has_consumer()) {
+                _runtime_bloom_filter_build_params.emplace_back(false, nullptr, kHashJoinKeyColumnOffset, -1);
```

Comment:
Why need pass kHashJoinKeyColumnOffset?

Reply:
in non-pipeline engine, kHashJoinKeyColumnOffset is pass from outerside.